src/util/futurepool.rs

```diff
@@ -0,0 +1,259 @@
+        self.inner.borrow_mut()
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.get_context().on_tick();
+    }
+}
+
+/// Each `ContextDelegator` instance is invoked individually for each thread
+/// so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextDelegator<T> {}
+
+#[derive(Debug)]
+pub struct ContextDelegators<T: Context> {
+    delegators: Arc<HashMap<thread::ThreadId, ContextDelegator<T>>>,
+}
+
+impl<T: Context> Clone for ContextDelegators<T> {
+    fn clone(&self) -> Self {
+        ContextDelegators::<T> {
+            delegators: Arc::clone(&self.delegators),
+        }
+    }
+}
+
+impl<T: Context> ContextDelegators<T> {
+    fn new(delegators: HashMap<thread::ThreadId, ContextDelegator<T>>) -> Self {
+        ContextDelegators::<T> {
+            delegators: Arc::new(delegators),
+        }
+    }
+
+    fn get_current_thread_delegator(&self) -> &ContextDelegator<T> {
+        let thread_id = thread::current().id();
+        if let Some(delegator) = self.delegators.get(&thread_id) {
+            delegator
+        } else {
+            unreachable!();
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A future thread pool that supports `on_tick` for each thread.
+pub struct FuturePool<T: Context> {
+    pool: cpupool::CpuPool,
+    context_delegators: ContextDelegators<T>,
+}
+
+impl<T: Context> Clone for FuturePool<T> {
+    fn clone(&self) -> FuturePool<T> {
+        FuturePool::<T> {
+            pool: self.pool.clone(),
+            context_delegators: self.context_delegators.clone(),
+        }
+    }
+}
+
+impl<T: Context> util::AssertSend for FuturePool<T> {}
+impl<T: Context> util::AssertSync for FuturePool<T> {}
+
+impl<T: Context> FuturePool<T> {
+    pub fn new<F>(
+        pool_size: usize,
+        stack_size: usize,
+        name_prefix: &str,
+        tick_interval: Duration,
+        context_factory: F,
+    ) -> FuturePool<T>
+    where
+        F: Send + 'static + Fn(thread::ThreadId) -> T,
+    {
+        let (tx, rx) = mpsc::sync_channel(pool_size);
+        let pool = cpupool::Builder::new()
+            .pool_size(pool_size)
+            .stack_size(stack_size)
+            .name_prefix(name_prefix)
+            .after_start(move || {
+                // We only need to know each thread's id and we can build context later
+                // by invoking `context_factory` in a !Sync way.
+                let thread_id = thread::current().id();
+                tx.send(thread_id).unwrap();
+            })
+            .create();
+        let contexts = (0..pool_size)
+            .map(|_| {
```

Comment:
You can iter `rx` directly.

Reply:
Previously it was. Changed to `(0..pool_size).map(|_| rx.recv())` according to @zhangjinpeng1987 's review to make it explicitly indicating that it will receive `pool_size` messages.

Think about this: what if only some of the `after_start` is called when executed to this statement? Will it only builds part of the contexts? Ok, according to the doc, `rx.iter` will block until its sender is closed.. then when will the sender close? Now we need to look at the `cpupool::Builder`'s `after_start` implementation to see when our `tx` is dropped. This is far less clear then current.

Reply:
See #[doc of Builder](http://alexcrichton.com/futures-rs/futures_cpupool/struct.Builder.html#method.after_start).

The `after_start` closure will be deconstructed after all worker threads call it. So, I think we can use a `Builder` to simplify code.

Reply:
I think it is too complicated to understand why `rx.into_iter().map....` is correct. We need to look at `Builder`'s doc as well as `Receiver`'s doc to know what will happen at all. The code is indeed simplified (a little) but its correctness becomes much harder to verify.

Reply:
@zhangjinpeng1987 What do you think